server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -398,6 +400,7 @@ public IndexShard(
         this.refreshPendingLocationListener = new RefreshPendingLocationListener();
         this.relativeTimeInNanosSupplier = relativeTimeInNanosSupplier;
         this.indexCommitListener = indexCommitListener;
+        this.fieldHasValue = new HashSet<>();
```

Comment:
Do we need to care about thread safety of this Set? I think we do not need, since we are only adding to it and never removing.

Reply:
++